etcdserver/v3_server.go

```diff
@@ -803,3 +804,105 @@ func (s *EtcdServer) AuthInfoFromCtx(ctx context.Context) (*auth.AuthInfo, error
 	return authInfo, nil
 
 }
+
+func (s *EtcdServer) Downgrade(ctx context.Context, r *pb.DowngradeRequest) (*pb.DowngradeResponse, error) {
+	switch r.Action {
+	case pb.DowngradeRequest_VALIDATE:
+		return s.downgradeValidate(ctx, r.Version)
+	case pb.DowngradeRequest_ENABLE:
+		return s.downgradeEnable(ctx, r)
+	case pb.DowngradeRequest_CANCEL:
+		return s.downgradeCancel(ctx)
+	default:
+		return nil, ErrUnknownDowngradeAction
+	}
+}
+
+func (s *EtcdServer) downgradeValidate(ctx context.Context, v string) (*pb.DowngradeResponse, error) {
+	resp := &pb.DowngradeResponse{}
+
+	targetVersion, err := changeToTargetVersion(v)
+	if err != nil {
+		return nil, err
+	}
+
+	// do linearized read to avoid using stale downgrade information
+	err = s.linearizableReadNotify(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	cv := s.ClusterVersion()
+	if cv == nil {
+		return nil, ErrClusterVersionUnavailable
+	}
+	resp.Version = cv.String()
+
+	allowedTargetVersion := semver.Version{Major: cv.Major, Minor: cv.Minor - 1}
+	if !targetVersion.Equal(allowedTargetVersion) {
+		err = ErrInvalidDowngradeTargetVersion
+		return nil, err
+	}
+
+	downgradeInfo := s.cluster.DowngradeInfo()
+
+	if downgradeInfo.Enabled {
+		// Todo: return the downgrade status along with the error msg
+		return resp, ErrIsDowngrading
+	}
+	return resp, nil
+}
+
+func (s *EtcdServer) downgradeEnable(ctx context.Context, r *pb.DowngradeRequest) (*pb.DowngradeResponse, error) {
+	// validate downgrade capability before starting downgrade
+	v := r.Version
+	if resp, err := s.downgradeValidate(ctx, v); err != nil {
+		return resp, err
+	}
+	targetVersion, err := changeToTargetVersion(v)
+	if err != nil {
+		return nil, err
+	}
+
+	raftRequest := membershippb.DowngradeInfoSetRequest{Enabled: true, Ver: targetVersion.String()}
+	_, err = s.raftRequest(ctx, pb.InternalRaftRequest{DowngradeInfoSet: &raftRequest})
+	if err != nil {
+		return nil, err
+	}
+	resp := pb.DowngradeResponse{Version: s.ClusterVersion().String()}
+	return &resp, nil
+}
+
+func (s *EtcdServer) downgradeCancel(ctx context.Context) (*pb.DowngradeResponse, error) {
+	// do linearized read to avoid using stale downgrade information
+	if err := s.linearizableReadNotify(ctx); err != nil {
+		return nil, err
+	}
+
+	downgradeInfo := s.cluster.DowngradeInfo()
+	if !downgradeInfo.Enabled {
+		return nil, ErrIsNotDowngrading
+	}
+
+	raftRequest := membershippb.DowngradeInfoSetRequest{Enabled: false}
```

Comment:
nit: If we set target version to 0 here, in apply stage we could simply reconstruct a new downgradeInfo from the request and overwrite the existing one.